Propagate mouse events to site-isolated frames

charliewolfe opened the pull request:
#### 31eb0baafdee9f427762f3fe7fbd8049663c6882
<pre>
Propagate mouse events to site-isolated frames
<a href="https://bugs.webkit.org/show_bug.cgi?id=260875">https://bugs.webkit.org/show_bug.cgi?id=260875</a>
rdar://114656583

Reviewed by Alex Christensen.

This patch makes changes to propagate mouse down, up, and move events to site-isolated frames.

When a hit test leads to a remote frame, the web process will send IPC to the UI process
containing the target frame ID and the transformed coordinates relative to the remote frame’s
parent. The UI process then sends the mouse event with the corrected coordinates to the specified
frame.

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::subframeForHitTestResult):
(WebCore::EventHandler::subframeForTargetNode):

Change these functions to return a `Frame` so we can later check if the target subframe is a
remote frame.

(WebCore::documentPointForWindowPoint):

`documentPointForWindowPoint()` should take a Frame so we can use it to transform mouse coordinates
to be sent to a remote frame.

(WebCore::EventHandler::handleMousePressEvent):
(WebCore::EventHandler::handleMouseDoubleClickEvent):
(WebCore::EventHandler::mouseMoved):
(WebCore::EventHandler::passMouseMovedEventToScrollbars):
(WebCore::EventHandler::handleMouseMoveEvent):
(WebCore::EventHandler::handleMouseReleaseEvent):

Update these functions to check when a remote frame is the target frame. If so, return
`RemoteMouseEventData`.

* Source/WebCore/page/HandleMouseEventResult.h: Added.
(WebCore::HandleMouseEventResult::HandleMouseEventResult):
(WebCore::HandleMouseEventResult::wasHandled):
(WebCore::HandleMouseEventResult::setHandled):
(WebCore::HandleMouseEventResult::remoteMouseEventData):

Add `HandleMouseEventResult` to replace to bool used as the return type for mouse events.

* Source/WebCore/page/FrameView.h:
* Source/WebCore/page/LocalFrameView.h:
* Source/WebCore/page/RemoteFrameView.h:

Create a virtual `frame()` function on FrameView.

* Source/WebCore/page/RemoteMouseEventData.h: Added.

Add `RemoteMouseEventData` to store the target remote frame and the transformed coordinates that is
sent to the UI process.

* Source/WebKit/Shared/WebMouseEvent.h:
(WebKit::WebMouseEvent::setPosition):

Add a `setPosition()` function that is used to correct the position of the propagated mouse event.

* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::remotePageProxy):
* Source/WebKit/UIProcess/WebFrameProxy.h:

Add an accessor for a RemotePageProxy.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::handleMouseEventReply):
(WebKit::WebPageProxy::sendMouseEvent):
(WebKit::WebPageProxy::processNextQueuedMouseEvent):
* Source/WebKit/UIProcess/WebPageProxy.h:

Add functions to propagate the mouse event to the frame specified in the completion handler in
`RemoteMouseEventData`.

* Source/WebCore/page/AutoscrollController.cpp:
(WebCore::AutoscrollController::stopAutoscrollTimer):
* Source/WebCore/page/ContextMenuController.cpp:
* Source/WebCore/page/ios/EventHandlerIOS.mm:
(WebCore::EventHandler::mouseDown):
(WebCore::EventHandler::mouseUp):
(WebCore::EventHandler::mouseMoved):
(WebCore::EventHandler::tryToBeginDragAtPoint):
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::handleMouseEvent):
* Source/WebKit/WebProcess/WebPage/WebFrame.h:
(WebKit::WebPage::mouseEvent):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::completeSyntheticClick):

Update these to use the new return type on the handle mouse event functions.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::TEST):

Add an API test to verify mouse up, down, and move events are sent to cross-origin frames when
site isolation is enabled. Also check that the mouse coordinates are correctly transformed.

Canonical link: <a href="https://commits.webkit.org/267964@main">https://commits.webkit.org/267964@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9ec3b57da3528f152e627ccdc49d7414b32c9652

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/18179 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/18510 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/19091 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/20018 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/17020 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/18376 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/21808 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/18668 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/18976 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/18400 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/18628 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/15829 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/20903 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/15852 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/16582 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/23091 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/16871 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/16753 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/20973 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/17325 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/14684 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/16412 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4336 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/20777 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/17170 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
```diff
@@ -1242,4 +1242,48 @@ HTTPServer server({
+    [navigationDelegate waitForDidFinishNavigation];
+
+    CGPoint eventLocationInWindow = [webView convertPoint:CGPointMake(50, 50) toView:nil];
+    [webView mouseEnterAtPoint:eventLocationInWindow];
```
This is only available #if PLATFORM(MAC)

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -12907,6 +12909,29 @@ void WebPageProxy::useRedirectionForCurrentNavigation(const ResourceResponse& re
+    if (!frame)
+        return;
+
+    auto remotePageProxy = internals().domainToRemotePageProxyMap.get(RegistrableDomain(frame->url()));
```
Make an accessor for WebFrameProxy's m_remotePageProxy and use that instead.

achristensen07 commented on Source/WebCore/page/LocalFrameView.h:
```diff
@@ -29,6 +29,7 @@
 #include "FrameView.h"
 #include "LayoutMilestone.h"
 #include "LayoutRect.h"
+#include "LocalFrame.h"
```
This shouldn't be necessary.

achristensen07 commented on Source/WebCore/page/RemoteFrameView.h:
```diff
@@ -26,19 +26,17 @@
 #pragma once
 
 #include "FrameView.h"
+#include "RemoteFrame.h"
```
This shouldn't be necessary.  You can put the definition of frame() into the cpp file.

achristensen07 commented on Source/WebCore/page/HandleMouseEventResult.h:
```diff
@@ -0,0 +1,51 @@
+
+struct HandleMouseEventResult {
+    bool handled;
+    std::optional<FrameIdentifier> targetFrameID;
```
Can we use Expected<bool, FrameIdentifier> instead?

achristensen07 commented on Source/WebCore/page/HandleMouseEventResult.h:
```diff
@@ -0,0 +1,51 @@
+
+    HandleMouseEventResult() = default;
+
+    HandleMouseEventResult(bool handled)
```
These explicit constructors seem excessive.

achristensen07 commented on Source/WebCore/page/ios/EventHandlerIOS.mm:
```diff
@@ -478,7 +478,7 @@ static bool findViewInSubviews(NSView *superview, NSView *target)
     CurrentEventScope scope(event);
 
-    event.wasHandled = handleMousePressEvent(currentPlatformMouseEvent());
+    event.wasHandled = handleMousePressEvent(currentPlatformMouseEvent()).value_or(false);
```
I think this should look more like this:
event.wasHandled = handleMousePressEvent(currentPlatformMouseEvent()).wasHandled();

Then the result class should use Expected<bool, something> as storage and have a bool wasHandled() method that returns the bool or false.  That'll make the code more readable and make future coders not wonder if they should use .value_or(false) or .value_or(true) and why.

charliewolfe commented on Source/WebCore/WebCore.xcodeproj/project.pbxproj:
```diff
@@ -6099,7 +6101,6 @@
 		F429261825264D4400007898 /* MediaPlayerIdentifier.h in Headers */ = {isa = PBXBuildFile; fileRef = F429261725264D4400007898 /* MediaPlayerIdentifier.h */; settings = {ATTRIBUTES = (Private, ); }; };
 		F429950F270E75000032298B /* DisplayListRecorder.h in Headers */ = {isa = PBXBuildFile; fileRef = F429950D270E75000032298B /* DisplayListRecorder.h */; settings = {ATTRIBUTES = (Private, ); }; };
 		F43160C82A2974D90021441D /* RunLoopObserver.h in Headers */ = {isa = PBXBuildFile; fileRef = F43160C72A2974D90021441D /* RunLoopObserver.h */; settings = {ATTRIBUTES = (Private, ); }; };
-		2D77ED552A9F563A000D0803 /* CryptoAlgorithmX25519Cocoa.cpp */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.cpp.cpp; path = CryptoAlgorithmX25519Cocoa.cpp; sourceTree = "<group>"; };
```
These moved somehow, will fix.

achristensen07 commented on Source/WebCore/WebCore.xcodeproj/project.pbxproj:
```diff
@@ -9025,10 +9028,10 @@
-		2D673E492A9CF2B100E2B4F0 /* CryptoAlgorithmX25519.h */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.c.h; path = CryptoAlgorithmX25519.h; sourceTree = "<group>"; };
 		2D5C9CFD19C7B52E00B3C5C1 /* PageOverlayController.cpp */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.cpp.cpp; path = PageOverlayController.cpp; sourceTree = "<group>"; };
 		2D5C9CFE19C7B52E00B3C5C1 /* PageOverlayController.h */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.c.h; path = PageOverlayController.h; sourceTree = "<group>"; };
+		2D673E482A9CF2B000E2B4F0 /* CryptoAlgorithmX25519.cpp */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.cpp.cpp; path = CryptoAlgorithmX25519.cpp; sourceTree = "<group>"; };
```
These too.  run sort-Xcode-project-file on this file.

achristensen07 commented on Source/WebCore/page/RemoteMouseEventData.h:
```diff
@@ -0,0 +1,59 @@
+
+    template<class Encoder> void encode(Encoder& encoder) const
+    {
+        encoder << targetFrameID;
```
Please put this in WebCoreArgumentCoders.serialization.in instead.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -3269,7 +3270,7 @@ void WebPageProxy::processNextQueuedMouseEvent()
     }
 
-    std::optional<Vector<SandboxExtension::Handle>> sandboxExtensions;
+    std::optional<Vector<SandboxExtension::Handle>> sandboxExtensions = std::nullopt;
```
= std::nullopt is unnecessary and redundant with the default std::optional constructor.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -3279,11 +3280,30 @@ void WebPageProxy::processNextQueuedMouseEvent()
+
+            auto remotePageProxy = frame->remotePageProxy();
+            if (!remotePageProxy)
+                return;
```
This can happen if a.com has an iframe at b.com which has an iframe at a.com and you click on the grandchild iframe.  You'd get a WebFrameProxy that doesn't have a RemotePageProxy because its WebPage communicates directly with the WebPageProxy.  In this case, we'd want to propagate the event and sandbox extensions back to the WebPageProxy.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -3279,11 +3280,30 @@ void WebPageProxy::processNextQueuedMouseEvent()
     LOG_WITH_STREAM(MouseHandling, stream << "UIProcess: sent mouse event " << eventType << " (queue size " << internals().mouseEventQueue.size() << ")");
     m_process->recordUserGestureAuthorizationToken(event.authorizationToken());
-    sendWithAsyncReply(Messages::WebPage::MouseEvent(m_mainFrame->frameID(), event, sandboxExtensions), [this, protectedThis = Ref { *this }] (WebEventType eventType, bool handled) {
+    CompletionHandler<void(WebEventType, bool, std::optional<RemoteMouseEventData>)> completionHandler = [this, protectedThis = Ref { *this }, sandboxExtensions = WTFMove(sandboxExtensions), &completionHandler] (WebEventType eventType, bool handled, std::optional<WebCore::RemoteMouseEventData> remoteMouseEventData) {
```
Yeah, this isn't quite right.  When this is called, from inside the lambda completionHandler will point to bad stack memory.  But this is a way to get a recursive lambda if you know it's going to be called immediately.  I do something similar in iso2022JPEncode but it turns out I don't need 2 lines!

I think what we want instead is a function on RemotePageProxy and WebPageProxy that takes a std::optional<Vector<SandboxExtension::Handle>> and event information and no completion handler.  It calls sendWithAsyncReply, and in the completion handler it checks for a RemoteMouseEventData and, if present, it propagates the event and sandbox extensions to the WebPageProxy or RemotePageProxy.  Also, organize it to make sure we don't have duplicate code.

achristensen07 commented on Source/WebCore/page/HandleMouseEventResult.h:
```diff
@@ -0,0 +1,59 @@
+    void setHandled(bool handled)
+    {
+        if (m_result.has_value())
+            m_result.value() = handled;
```
.value() is probably not needed

achristensen07 commented on Source/WebCore/page/HandleMouseEventResult.h:
```diff
@@ -0,0 +1,59 @@
+    {
+    }
+
+    bool wasHandled() { return m_result.has_value() ? m_result.value() : false; }
```
nit: I tend to like expected's operator bool and operator*.  This is fine, but I think it could also be abbreviated to "return m_result ? *m_result : false;"

achristensen07 commented on Source/WebCore/page/HandleMouseEventResult.h:
```diff
@@ -0,0 +1,59 @@
+
+    std::optional<RemoteMouseEventData> remoteMouseEventData()
+    {
+        return m_result.has_value() ? std::nullopt : std::optional<RemoteMouseEventData>(m_result.error());
```
nit: .has_value() is probably not needed.

achristensen07 reviewed (approved):
Hooray!

charliewolfe committed:
Propagate mouse events to site-isolated frames
https://bugs.webkit.org/show_bug.cgi?id=260875
rdar://114656583

Reviewed by Alex Christensen.

This patch makes changes to propagate mouse down, up, and move events to site-isolated frames.

When a hit test leads to a remote frame, the web process will send IPC to the UI process
containing the target frame ID and the transformed coordinates relative to the remote frame’s
parent. The UI process then sends the mouse event with the corrected coordinates to the specified
frame.

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::subframeForHitTestResult):
(WebCore::EventHandler::subframeForTargetNode):

Change these functions to return a `Frame` so we can later check if the target subframe is a
remote frame.

(WebCore::documentPointForWindowPoint):

`documentPointForWindowPoint()` should take a Frame so we can use it to transform mouse coordinates
to be sent to a remote frame.

(WebCore::EventHandler::handleMousePressEvent):
(WebCore::EventHandler::handleMouseDoubleClickEvent):
(WebCore::EventHandler::mouseMoved):
(WebCore::EventHandler::passMouseMovedEventToScrollbars):
(WebCore::EventHandler::handleMouseMoveEvent):
(WebCore::EventHandler::handleMouseReleaseEvent):

Update these functions to check when a remote frame is the target frame. If so, return
`RemoteMouseEventData`.

* Source/WebCore/page/HandleMouseEventResult.h: Added.
(WebCore::HandleMouseEventResult::HandleMouseEventResult):
(WebCore::HandleMouseEventResult::wasHandled):
(WebCore::HandleMouseEventResult::setHandled):
(WebCore::HandleMouseEventResult::remoteMouseEventData):

Add `HandleMouseEventResult` to replace to bool used as the return type for mouse events.

* Source/WebCore/page/FrameView.h:
* Source/WebCore/page/LocalFrameView.h:
* Source/WebCore/page/RemoteFrameView.h:

Create a virtual `frame()` function on FrameView.

* Source/WebCore/page/RemoteMouseEventData.h: Added.

Add `RemoteMouseEventData` to store the target remote frame and the transformed coordinates that is
sent to the UI process.

* Source/WebKit/Shared/WebMouseEvent.h:
(WebKit::WebMouseEvent::setPosition):

Add a `setPosition()` function that is used to correct the position of the propagated mouse event.

* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::remotePageProxy):
* Source/WebKit/UIProcess/WebFrameProxy.h:

Add an accessor for a RemotePageProxy.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::handleMouseEventReply):
(WebKit::WebPageProxy::sendMouseEvent):
(WebKit::WebPageProxy::processNextQueuedMouseEvent):
* Source/WebKit/UIProcess/WebPageProxy.h:

Add functions to propagate the mouse event to the frame specified in the completion handler in
`RemoteMouseEventData`.

* Source/WebCore/page/AutoscrollController.cpp:
(WebCore::AutoscrollController::stopAutoscrollTimer):
* Source/WebCore/page/ContextMenuController.cpp:
* Source/WebCore/page/ios/EventHandlerIOS.mm:
(WebCore::EventHandler::mouseDown):
(WebCore::EventHandler::mouseUp):
(WebCore::EventHandler::mouseMoved):
(WebCore::EventHandler::tryToBeginDragAtPoint):
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::handleMouseEvent):
* Source/WebKit/WebProcess/WebPage/WebFrame.h:
(WebKit::WebPage::mouseEvent):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::completeSyntheticClick):

Update these to use the new return type on the handle mouse event functions.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::TEST):

Add an API test to verify mouse up, down, and move events are sent to cross-origin frames when
site isolation is enabled. Also check that the mouse coordinates are correctly transformed.

Canonical link: https://commits.webkit.org/267964@main
